CARDS-853: Transfer filters from Dashboard to forms page when clicking on Expand

veronikaslc opened the pull request.

Commit:
CARDS-853: Transfer filters from Dashboard to forms page when clicking on Expand

marta- commented:
**Bug 1**: The filters are transferred but not applied, see screencast for details.

**Bug 2**: After the filters are transferred, clicking on (+) to add more filters results in a React white screen. Several warnings are logged, and the following error: `Uncaught TypeError: Cannot read property 'dataType' of undefined`.

![CPT2109010933-1637x491](https://user-images.githubusercontent.com/651980/131679981-c2a6c18c-ea53-476b-b1df-77597918d647.gif)

Commit:
CARDS-853: Transfer filters from Dashboard to forms page when clicking on Expand
Fixed filter transfering

marta- reviewed (approved):
Looks good to me

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/FormView.jsx:
```diff
@@ -49,6 +49,7 @@ function FormView(props) {
   const [ subtitle, setSubtitle ] = useState();
   const [ questionnairePath, setQuestionnairePath ] = useState();
   const [ qFetchSent, setQFetchStatus ] = useState(false);
+  const [ filtersJsonString, setFiltersJsonString ] = useState(expanded ? location.hash.split(/&filters=(.+)/)[1] || "" : "");
```
This is fragile, we may add other things after filters in the future. To be more flexible, I would use something like:

```suggestion
  const [ filtersJsonString, setFiltersJsonString ] = useState(expanded ? new URLSearchParams(window.location.hash.substring(1)).get("filters") : "");
```

But we could go further, and use different parameter names for different filters. Here we could identify that these filters apply to the forms table with:

```suggestion
  const [ filtersJsonString, setFiltersJsonString ] = useState(expanded ? new URLSearchParams(window.location.hash.substring(1)).get("forms:filters") : "");
```

And we could also apply the filters when loading the dashboard, to allow sharing a link to the dashboard with the current filters applied:

```suggestion
  const [ filtersJsonString, setFiltersJsonString ] = useState(new URLSearchParams(window.location.hash.substring(1)).get("forms:filters"));
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/FormView.jsx:
```diff
@@ -75,7 +76,7 @@ function FormView(props) {
 
   const tabs = ["Completed", "Draft"];
-  const activeTabParam = location.hash.substr(1);
+  const activeTabParam = location.hash.substr(1).split('&')[0];
```
We should refactor how we identify the active tab as well:

```suggestion
  const activeTabParam = new URLSearchParams(window.location.hash.substring(1)).get("forms:activeTab");
```

Of course, it's also needed to update the part where the URL is generated.

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/FormView.jsx:
```diff
@@ -114,7 +115,7 @@ function FormView(props) {
           !expanded &&
           <Tooltip title="Expand">
-            <Link to={"/content.html/Forms#" + tabs[activeTab]}>
+            <Link to={"/content.html/Forms#" + tabs[activeTab] + (filtersJsonString ? "&filters=" + filtersJsonString : "")}>
```
```suggestion
            <Link to={"/content.html/Forms#" + new URLSearchParams({"forms:activeTab" : tabs[activeTab], "forms:filters" : filtersJsonString}).toString()}>
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/FormView.jsx:
```diff
@@ -140,6 +141,8 @@ function FormView(props) {
           actions={actions}
           disableTopPagination={!topPagination}
+          onFiltersChange={(str) => { !expanded && setFiltersJsonString(str); }}
+          filtersJsonString={expanded ? filtersJsonString : ""}
```
```suggestion
          onFiltersChange={(str) => { setFiltersJsonString(str); }}
          filtersJsonString={filtersJsonString}
```

sdumitriu committed:
CARDS-853: Transfer filters from Dashboard to forms page when clicking on Expand
- allow reading filters and active tab from the URL on the dashboard too
- allow settings for both forms and subjects in the URL
- no more special case for the active tab, this setting also has a name in the URL

sashaandjic reviewed (approved):
re-tested
